### PR TITLE
8279833: Loop optimization issue in String.encodeUTF8_UTF16

### PR DESCRIPTION
**Pull request by cl4es**
In `String.encodeUTF8_UTF16`, making the `char c` local to each loop helps the performance of the method by helping C2 optimize each individual loop better.

Results on the updated micros:
19-b04:
```
Benchmark                          (charsetName)  Mode  Cnt     Score     Error  Units
StringEncode.encodeUTF16                   UTF-8  avgt   15   164.457 ±   7.296  ns/op
StringEncode.encodeUTF16LongEnd            UTF-8  avgt   15  2294.160 ±  40.580  ns/op
StringEncode.encodeUTF16LongStart          UTF-8  avgt   15  9128.698 ± 124.636  ns/op
```

Patch:
```
Benchmark                          (charsetName)  Mode  Cnt     Score    Error  Units
StringEncode.encodeUTF16                   UTF-8  avgt   15   131.296 ±  6.693  ns/op
StringEncode.encodeUTF16LongEnd            UTF-8  avgt   15  2282.750 ± 46.891  ns/op
StringEncode.encodeUTF16LongStart          UTF-8  avgt   15  4786.965 ± 64.896  ns/op
```

Going back this seem to have been an issue since this method was introduced with JEP 254 in JDK 9:

8u311:
```
Benchmark                          (charsetName)  Mode  Cnt     Score     Error  Units
StringEncode.encodeUTF16                   UTF-8  avgt   15   194.057 ±   3.913  ns/op
StringEncode.encodeUTF16LongEnd            UTF-8  avgt   15  3024.860 ±  88.386  ns/op
StringEncode.encodeUTF16LongStart          UTF-8  avgt   15  5282.849 ± 247.230  ns/op
```
9:
```
Benchmark                          (charsetName)  Mode  Cnt      Score     Error  Units
StringEncode.encodeUTF16                   UTF-8  avgt   15    148.481 ±   9.374  ns/op
StringEncode.encodeUTF16LongEnd            UTF-8  avgt   15   2832.754 ± 263.372  ns/op
StringEncode.encodeUTF16LongStart          UTF-8  avgt   15  10447.115 ± 408.338  ns/op
```

(Interestingly JDK 9 seem faster on some of the micros than later iterations, while slower on others. The main issue is the slow non-ASCII loop, where the patch speeds things up ~2x. With the patch we're significantly faster than both JDK 8 and 9 on all measures.)

There's a JIT compiler bug hiding in plain sight here where the potentially uninitialized local `char c` appears to mess up optimization of the second loop. I'll file a separate bug for this (a standalone reproducer should be straightforward to produce). I think this patch is reasonable to put back into the JDK while we investigate if/how C2 can better handle this kind of condition. It might also be easier to backport, depending on whether the C2 fix is trivial or not.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279833](https://bugs.openjdk.java.net/browse/JDK-8279833): Loop optimization issue in String.encodeUTF8_UTF16

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 70d0c5c6e04622f8e806debae650c704a1d24d5e
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7026/head:pull/7026` \
`$ git checkout pull/7026`

Update a local copy of the PR: \
`$ git checkout pull/7026` \
`$ git pull https://git.openjdk.java.net/jdk pull/7026/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7026`

View PR using the GUI difftool: \
`$ git pr show -t 7026`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7026.diff">https://git.openjdk.java.net/jdk/pull/7026.diff</a>

</details>

### CODE REVIEW

File: test/micro/org/openjdk/bench/java/lang/StringEncode.java [hunk 1]
```diff
@@ -30,59 +30,97 @@
...
+    @Benchmark
+    public void encodeMix(Blackhole bh) throws Exception {
+        bh.consume(asciiString.getBytes(charset));
+        bh.consume(utf16String.getBytes(charset));
     }
 
-    private String asciiDefaultString;
-    private String utf16DefaultString;
+    @Benchmark
+    public byte[] encodeUTF16LongEnd(Blackhole bh) throws Exception {
```
Comment (shipilev) [inline test/micro/org/openjdk/bench/java/lang/StringEncode.java:113]: Here and later: can drop `bh` argument, since you don't need it anymore.

**Review by shipilev — APPROVED** [review]
Looks fine, consider touchups in benchmark code.

**Review by AlanBateman — APPROVED** [review]
A good find, hopefully it will be linked to the C2 issue.

This is the first change to both files in 2022 so you might have to update the copyright headers.

**Review by AlanBateman — APPROVED** [review]

**Comment by cl4es** [comment]
Thanks for reviewing, @shipilev and @AlanBateman

/integrate

**Comment by stsypanov** [comment]
Looks like there's one more similar case: https://github.com/openjdk/jdk/pull/7034

**Pull request closed without merging**